Supported Living Payment entitlement & benefit rates

jadedarko opened the pull request:
* Tax and benefit system evolution
* Impacted periods: all
* Impacted areas:
  * `openfisca_aotearoa/parameters/social_security/supported_living_payment/base.yaml`
  * `openfisca_aotearoa/variables/acts/social_security/interpretation/child.py`
  * `openfisca_aotearoa/variables/acts/social_security/schedule_4/part3.py`
  * `openfisca_aotearoa/variables/acts/social_security/supported_living_payment/supported_living_payment.py`
  * `openfisca_aotearoa/variables/acts/social_security/supported_living_payment/supported_living_payment__benefit.py`
* Details:
  - Added 2018 version for Supported Living Payment
  - Added differential carer age requirements for both legislation versions
  - Reorganised functions to better match legislation

- - - -

These changes:

- Impact the OpenFisca-Aotearoa public API (for instance renaming or removing a variable)
- Add a new feature (for instance adding a variable)
- Fix or improve an already existing calculation.
- Change non-functional parts of this repository (for instance editing the README)

jadedarko commented:
@verbman @Br3nda might either of you have time to look over this?

verbman commented on openfisca_aotearoa/parameters/social_security/supported_living_payment/base.yaml:
```diff
@@ -0,0 +1,827 @@
+description: Supported Living Payment cut-out points
```
Thank you so much for the research gone into this parameter set, it looks perfect!

verbman commented on openfisca_aotearoa/variables/acts/social_security/interpretation/child.py:
```diff
@@ -41,6 +41,7 @@ def formula(persons, period, parameters):
 
 
 class social_security__dependent_children(variables.Variable):
+    default_value = 0
```
Agreed!

verbman commented on openfisca_aotearoa/variables/acts/social_security/schedule_4/part3.py:
```diff
@@ -0,0 +1,324 @@
+    value_type = bool
+    entity = entities.Person
+    definition_period = periods.WEEK
+    reference = "https://www.legislation.govt.nz/act/public/2018/0032/latest/DLM6784861.html"
```
I'd suggest we also include a reference to the original version of the 2018 act so it's easier to track down the original formula and perhaps a link as well to the date it changed:
https://www.legislation.govt.nz/act/public/2018/0032/77.0/DLM6784861.html
https://www.legislation.govt.nz/act/public/2018/0032/140.0/DLM6784861.html

verbman commented on openfisca_aotearoa/variables/acts/social_security/schedule_4/part3.py:
```diff
@@ -0,0 +1,324 @@
+    value_type = bool
+    entity = entities.Person
+    definition_period = periods.WEEK
+    reference = "https://www.legislation.govt.nz/act/public/2018/0032/latest/DLM6784861.html"
```
As above - I'd suggest we also include a reference to the original version of the 2018 act so it's easier to track down the original formula and perhaps a link as well to the date it changed:
https://www.legislation.govt.nz/act/public/2018/0032/77.0/DLM6784861.html
https://www.legislation.govt.nz/act/public/2018/0032/140.0/DLM6784861.html

verbman commented on openfisca_aotearoa/variables/acts/social_security/schedule_4/part3.py:
```diff
@@ -0,0 +1,324 @@
+    value_type = bool
+    entity = entities.Person
+    definition_period = periods.WEEK
+    reference = "https://www.legislation.govt.nz/act/public/2018/0032/latest/DLM6784861.html"
```
As above:
https://www.legislation.govt.nz/act/public/2018/0032/77.0/DLM6784861.html
https://www.legislation.govt.nz/act/public/2018/0032/140.0/DLM6784861.html

verbman commented on openfisca_aotearoa/variables/acts/social_security/schedule_4/part3.py:
```diff
@@ -0,0 +1,324 @@
+    value_type = bool
+    entity = entities.Person
+    definition_period = periods.WEEK
+    label = """Shortcut to indicate beneficiary meets the following from clause 5:
```
Wrapping up concepts like this is quite common. I'd replace "Shortcut to indicate" to "Indicates".

verbman commented on openfisca_aotearoa/variables/acts/social_security/schedule_4/part3.py:
```diff
@@ -0,0 +1,324 @@
+    value_type = bool
+    entity = entities.Person
+    definition_period = periods.WEEK
+    label = """Shortcut to indicate beneficiary meets the following from clause 6:
```
Wrapping up concepts like this is quite common. I'd replace "Shortcut to indicate" to "Indicates".

verbman commented on openfisca_aotearoa/variables/acts/social_security/supported_living_payment/supported_living_payment.py:
```diff
@@ -1,124 +1,236 @@
+
+        return ssa34_a * ssa34_b * ssa34_c * ssa36
+
+    def formula(persons, period, parameters):
```
This 40B section appears to be referring to the 1964 Act. The formula definition possibly needs a date as it's not clear when this formula is required or that it's an older version. If the 1964 act logic is required (I'm not yet clear why it is) we should also had a link to it in the reference above based on the appropriate dated version

verbman replied:
Just to clarify, I really appreciate the 1964 Act being addressed but not sure it's a requirement right now.

verbman commented on openfisca_aotearoa/variables/acts/social_security/supported_living_payment/supported_living_payment.py:
```diff
@@ -1,124 +1,236 @@
+        "https://www.legislation.govt.nz/act/public/2018/0032/latest/whole.html#DLM6783174",
+        ]
 
+    def formula_2018_11_26(persons, period, parameters):
```
Just want to say how great it is to see the internal logic laid out like this, makes reviewing much easier!

verbman commented on openfisca_aotearoa/variables/acts/social_security/supported_living_payment/supported_living_payment.py:
```diff
@@ -1,124 +1,236 @@
+        return ssa34 + ssa40
+
+
+class supported_living_payment__disabled_or_blind__entitled(variables.Variable):
```
Adhering to the naming convention would mean dropping the "__entitled" suffix from this formula name.

verbman commented on openfisca_aotearoa/variables/acts/social_security/supported_living_payment/supported_living_payment.py:
```diff
@@ -1,124 +1,236 @@
+        return ssa40B_1 * ssa40B_5
+
+
+class supported_living_payment__carer__entitled(variables.Variable):
```
Adhering to the naming convention would mean dropping the "__entitled" suffix from this formula name.

verbman commented on openfisca_aotearoa/variables/acts/social_security/supported_living_payment/supported_living_payment.py:
```diff
@@ -1,124 +1,236 @@
-    set_input = holders.set_input_dispatch_by_period
+    label = "Eligible for Supported Living Payment, on grounds of caring for another person."
+    reference = [
+        "https://www.legislation.govt.nz/act/public/2018/0032/latest/whole.html#DLM6783187",
```
Add references to the relevant versions of the1964 Act:
https://www.legislation.govt.nz/act/public/1964/0136/219.0/DLM5469766.html
https://www.legislation.govt.nz/act/public/1964/0136/334.0/DLM5469766.html

verbman commented on openfisca_aotearoa/variables/acts/social_security/supported_living_payment/supported_living_payment.py:
```diff
@@ -1,124 +1,236 @@
     definition_period = periods.WEEK
     label = "Is incapable of regularly working 15 or more hours a week in open employment"
-    reference = "https://www.legislation.govt.nz/act/public/2018/0032/latest/whole.html#DLM6783176", "http://legislation.govt.nz/act/public/1964/0136/latest/whole.html#DLM5468367"
+    reference = [
```
Much prefer this format, it'd be good to add this to the Contributing documentation

verbman commented on openfisca_aotearoa/variables/acts/social_security/supported_living_payment/supported_living_payment__benefit.py:
```diff
@@ -0,0 +1,304 @@
+        principal_blind = population("totally_blind", period)
+        principal_disabled = population("supported_living_payment__restricted_work_capacity", period)
+        principal_abled = numpy.logical_not(numpy.logical_or(principal_blind, principal_disabled))
+        principal_disabled_income = gross_principal_income - (20 * period.size_in_weeks)
```
Should probably set up this $20 amount as a parameter

verbman commented on openfisca_aotearoa/variables/acts/social_security/supported_living_payment/supported_living_payment__benefit.py:
```diff
@@ -0,0 +1,304 @@
+    entity = entities.Person
+    definition_period = periods.WEEK
+    label = "The amount the base benefit is reduced base on the appropriate Income Test and the person & their partners income"
+    reference = "https://legislation.govt.nz/act/public/2018/0032/latest/whole.html#DLM6784861"
```
I wonder if we should update the contributing documentation to cover references in more detail. As a reviewer it would be very helpful for the reference to include each of the versions that inspire the formulas:
That is we use the "latest" link until such time as the logic is changed and then we switch to using the earliest versions the logic applies to, e.g. in this situation:
 https://www.legislation.govt.nz/act/public/2018/0032/56.0/DLM6784861.html (2018 11 26)
 https://www.legislation.govt.nz/act/public/2018/0032/140.0/DLM6784861.html (2020 11 09)

verbman commented on openfisca_aotearoa/variables/acts/social_security/supported_living_payment/supported_living_payment__benefit.py:
```diff
@@ -0,0 +1,304 @@
+    value_type = float
+    entity = entities.Person
+    definition_period = periods.WEEK
+    reference = "https://www.legislation.govt.nz/act/public/2018/0032/latest/DLM6784850.html"
```
As above:
https://www.legislation.govt.nz/act/public/2018/0032/56.0/DLM6784861.html
https://www.legislation.govt.nz/act/public/2018/0032/140.0/DLM6784861.html

jadedarko commented:
@verbman I believe I've now addressed the comments you made:

- reference lists contain all versions where multiple formula versions exist for a variable
- $20 amount has been moved to a parameter
- contributing documentation has been updated with references in list format & an explainer
- "shortcut to indicate" changed to "Indicates"
- redundant `__entitled` suffix removed
- 1964 version formulae removed due to not being complete across the board

verbman reviewed (approved):
Thanks!
